### PR TITLE
refactor: addition of rbac needed for instascale controller

### LINKED ISSUE #278: Add RBAC required by the InstaScale controller
## WHY

Part of the CodeFlare operator re-design, as described in project-codeflare/adr#9. 

## WHAT

Include the RBAC manifests required by the InstaScale controller, e.g., those required to consume the MachineSet and ClusterVersion APIs. 

## HOW

Copy the InstaScale RBAC into separate manifests.

To be done on top of #216.

## TESTS

Make sure the embedded InstaScale controller works as expected.

### PR DESCRIPTION
**Pull request by dimakis**
# Issue link

closes #278 

# What changes have been made

Addition of a ClusterRole and a ClusterRoleBinding which binds the required permissions to the CFO serviceaccount

# Verification steps

Run through demo instascale NBs with clusters leveraging machinepools (OSD) and machinesets (instructions for sample can be found [here](https://github.com/Maxusmusti/dev-clusters))

## Checks
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: config/rbac/instascale_role.yaml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  name: instascale-role
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - configmaps
+  verbs:
+  - get
+- apiGroups:
+  - ""
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:15]: It's already part of the "core" CusterRole, so it's preferable not duplicating it.

File: config/rbac/instascale_role.yaml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  verbs:
+  - get
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  verbs:
+  - get
+  - list
+  - watch
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:38]: It's already part of the main/MCAD CusterRole, so it's preferable not duplicating it.

File: config/rbac/instascale_role.yaml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  resources:
+  - clusterversions
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - machine.openshift.io
+  resources:
+  - '*'
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:68]: Could wildcard be avoided? It's generally a best practice recommended by SRE.

File: config/rbac/instascale_role.yaml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  - nodes
+  verbs:
+  - get
+  - list
+  - patch
+  - update
+- apiGroups:
+  - ""
+  resourceNames:
+  - instascale-ocm-secret
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:26]: The OCM Secret can now be referenced in the operator configuration, so it's not hard-coded anymore.

File: config/rbac/instascale_role.yaml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  verbs:
+  - update
+- apiGroups:
+  - workload.codeflare.dev
+  resources:
+  - appwrappers/status
+  verbs:
+  - get
+  - patch
+  - update
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:102]: It's already part of the main/MCAD CusterRole, so it's preferable not duplicating these.

**Review by astefanutti — CHANGES_REQUESTED** [review]

File: config/rbac/instascale_role.yaml [hunk 6]
```diff
@@ -65,19 +49,10 @@ rules:
...
-  - list
-  - patch
-  - update
-  - watch
-- apiGroups:
-  - workload.codeflare.dev
-  resources:
-  - appwrappers
+  - controlplanemachinesets
+  - machinehealthchecks
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:53]: Are these two resources really used by InstaScale?

Comment (dimakis) [reply config/rbac/instascale_role.yaml:53]: I should have specified I've pushed this as myself and mark are testing which are actually used and what we can cut away.

Comment (astefanutti) [reply config/rbac/instascale_role.yaml:53]: My understanding is that InstaScale only uses the Machine and MachineSet APIs but better double checking.

**Review by Bobbins228 — CHANGES_REQUESTED** [review]
Tested this on an OCP cluster with Antonin's requested changes and got expected InstaScale behaviour.

When testing with OSD I received this error on scale up:
```
E0925 15:58:54.827990 1 machinepools.go:58] Error creating MachinePool: expected response content type 'application/json' but received '' and content ''
I0925 15:58:54.828009 1 machinepools.go:60] Created MachinePool: <nil>
E0925 15:58:55.263876 1 machinepools.go:58] Error creating MachinePool: expected response content type 'application/json' but received '' and content ''
I0925 15:58:55.263893 1 machinepools.go:60] Created MachinePool: <nil>
```

**Comment by astefanutti** [comment]
> When testing with OSD I received this error on scale up:
> 
> ```
> E0925 15:58:54.827990 1 machinepools.go:58] Error creating MachinePool: expected response content type 'application/json' but received '' and content ''
> I0925 15:58:54.828009 1 machinepools.go:60] Created MachinePool: <nil>
> E0925 15:58:55.263876 1 machinepools.go:58] Error creating MachinePool: expected response content type 'application/json' but received '' and content ''
> I0925 15:58:55.263893 1 machinepools.go:60] Created MachinePool: <nil>
> ```

Could it be that we are missing permissions on `machinepools` or other related resources?

File: config/rbac/instascale_role.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - apps
+  resources:
+  - machineset/status
+  verbs:
+  - get
```
Comment (astefanutti) [inline config/rbac/instascale_role.yaml:40]: Are MachineSet APIs really in the `apps` group? It seems it's a wrong duplicate of the `machine.openshift.io` one.

Comment (dimakis) [reply config/rbac/instascale_role.yaml:40]: nope, I didn't think they were in the `apps` group, I thought first class objects like replicasets, deployments etc were.

but I was just copying what was in the [original role](https://github.com/project-codeflare/instascale/blob/2c7538eb8f93b77a768bd774bf963f37a0c41b23/config/rbac/role.yaml#L39-L56) to try get it working.

I can't really figure out why this is the case, even getting rid of those and allowing the wildcard on teh machine.openshift.io group doesn't returns the same error as ye got. 
I also can't seem to find a machinepool resource for any api group in any of the docs. 
Any ideas here?

Comment (astefanutti) [reply config/rbac/instascale_role.yaml:40]: Actually, I don't think the error is related to the RBAC changes. I suspect there is an issue with the OCM client configuration.

**Pull request merged**